lib/SILOptimizer/Transforms/DeadCodeElimination.cpp

```diff
@@ -323,32 +329,60 @@ void DCE::markLive() {
       }
       case SILInstructionKind::BeginBorrowInst: {
         auto *borrowInst = cast<BeginBorrowInst>(&I);
+        auto borrow = BorrowedValue(borrowInst);
         // Populate guaranteedPhiDependencies for this borrowInst
-        findGuaranteedPhiDependencies(BorrowedValue(borrowInst));
-        auto disableBorrowDCE = [&](SILValue borrow) {
-          visitTransitiveEndBorrows(borrow, [&](EndBorrowInst *endBorrow) {
-            markInstructionLive(endBorrow);
-          });
-        };
-        // If we have a begin_borrow of a @guaranteed operand, disable DCE'ing
-        // of parent borrow scopes. Dead reborrows needs complex handling, which
-        // is why it is disabled for now.
-        if (borrowInst->getOperand()->getOwnershipKind() ==
-            OwnershipKind::Guaranteed) {
-          SmallVector<SILValue, 4> roots;
-          findGuaranteedReferenceRoots(borrowInst->getOperand(),
-                                       /*lookThroughNestedBorrows=*/false,
-                                       roots);
-          // Visit the end_borrows of all the borrow scopes that this
-          // begin_borrow could be borrowing, and mark them live.
-          for (auto root : roots) {
-            disableBorrowDCE(root);
+        findGuaranteedPhiDependencies(borrow);
+        if (borrow.hasReborrow()) {
+          // If the load_borrow was computed from another
+          // borrow, disable DCE of the parent borrow.
+          // This is because when a reborrow is dead, DCE has to insert
+          // end_borrows in predecessor blocks and it cannot yet handle borrow
+          // nesting.
+          // TODO: Instead of disabling DCE of parent borrow, consider using
+          // lifetime completion to insert end_borrows.
+          if (borrowInst->getOperand()->getOwnershipKind() ==
+              OwnershipKind::Guaranteed) {
+            SmallVector<SILValue, 4> roots;
+            findGuaranteedReferenceRoots(borrowInst->getOperand(),
+                                         /*lookThroughNestedBorrows=*/false,
+                                         roots);
+            // Visit the end_borrows of all the borrow scopes that this
+            // begin_borrow could be borrowing, and mark them live.
+            for (auto root : roots) {
+              disableBorrowDCE(root);
+            }
           }
         }
         break;
       }
       case SILInstructionKind::LoadBorrowInst: {
-        findGuaranteedPhiDependencies(BorrowedValue(cast<LoadBorrowInst>(&I)));
+        auto *loadBorrowInst = cast<LoadBorrowInst>(&I);
```

Comment:
Can you avoid this code duplication? Either by extracting it into a helper function or even merging the two switch cases